src/Router.ts

```diff
@@ -0,0 +1,198 @@
+import MissingRouteError from './errors/MissingRouteError';
+import has from './has';
+import HashSource from './HashSource';
+import HtmlHistorySource from './HtmlHistorySource';
+import { CancelableNavigationArgs, NavigationArgs, RouterArgs, RouterSource } from './interfaces';
+import { joinPath, normalizePath } from './PathRule';
+import RouteGroup from './RouteGroup';
+
+const DefaultSource = has('html5-history') ? HtmlHistorySource : HashSource;
+let defaultSource: RouterSource = null;
+
+if (has('host-browser')) {
+	defaultSource = new (<any> DefaultSource)();
+}
+
+/**
+ * Handles navigating to/from paths in response to events from the user/environment.
+ *
+ * @fires "beforechange" Before the lifecycle methods are called. Passed an `event` object with route
+ * 		data and a `preventDefault` method that can be called to cancel the route change.
+ * @fires "change" When the new route has been entered. Passed an object containing route data.
+ * @fires "error" When a route change is canceled or another error occurs. Passed the error object.
+ */
+export default class Router extends RouteGroup {
+	protected _evented: Evented;
+
+	/**
+	 * The route data for a route canceled by one of the following methods:
+	 * `beforeEnter`, `beforeExit`, or a `beforechange` event handler. This is cleared
+	 * each time a new route is successfully entered.
+	 */
+	canceled: NavigationArgs;
+
+	/**
+	 * The data for the current route.
+	 */
+	current: NavigationArgs;
+
+	/**
+	 * The object responsible for updating the environment in response to router changes. Defaults to
+	 * an instance of `HtmlHistorySource` in environments where the history API is supported, and to
+	 * an instance of `HashSource` otherwise.
+	 *
+	 * `Router` does not actually change the browser URL or interact with the environment directly.
+	 * This allows users to control how the browser responds to route changes (for example, by changing
+	 * the URL hash or by calling `history.pushState`).
+	 */
+	source: RouterSource;
+
+	constructor(kwArgs: RouterArgs) {
+		super(kwArgs);
+
+		this.source = kwArgs.source || defaultSource;
+		this._evented = new Evented();
+	}
+
+	protected _createEvent(type: string, args: NavigationArgs): CancelableNavigationArgs {
+		const event = <CancelableNavigationArgs> assign(Object.create(null), args);
+		event.type = type;
+
+		return event;
+	}
+
+	protected _getEventPromise(
+		event: CancelableNavigationArgs,
+		path: string,
+		object: any,
+		method: string
+	): Promise<any> {
+		return new Promise(function (resolve: Function, reject: Function): void {
+			if (typeof object[method] !== 'function') {
+				return resolve();
+			}
+
+			event.preventDefault = function () {
+				reject(new CancelNavigationError('Navigation canceled for path ' + path));
+			};
+
+			const handle = after(object, method, function () {
+				handle.destroy();
+				resolve();
+			});
+			object[method](event);
```

Comment:
The approach is really good. Looking at the after callback again, I'm not sure I should be concerned.

My thought was if `object[method]()` throws then `handle.destroy()` won't be called which would keep a reference to the closure tied to the object and if `object[method]()` is called again then the attached after callback will be called on success.  I think that I'm a little paranoid because `_getEventPromise()` is very generic.

Anyway, I wanted to highlight that and if you're comfortable with it then this doesn't need to change.